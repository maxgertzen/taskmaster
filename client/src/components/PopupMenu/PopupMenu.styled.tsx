import styled from '@emotion/styled';

export const MenuContainer = styled('div')<{
  top: number;
}>(({ theme, top }) => ({
  position: 'absolute',
  top,
  right: 0,
  maxWidth: '100%',
  background: theme.colors.surface,
  border: theme.borders.main(theme.colors.grey),
  boxShadow: '0px 4px 6px rgba(0, 0, 0, 0.1)',
  borderRadius: theme.borders.radius,
  zIndex: 1000,
  display: 'flex',
  flexDirection: 'column',
  padding: theme.spacing(1, 0),
}));

export const MenuItem = styled.span<{ disabled?: boolean }>(
  ({ theme, disabled }) => ({
    all: 'unset',
    padding: theme.spacing(1, 2),
    cursor: disabled ? 'not-allowed' : 'pointer',
    textAlign: 'left',
    ...(!disabled && {
      '&:hover': {
        background: theme.colors.grey,
        color: theme.colors.secondary,
      },
      '&:focus': {
        background: theme.colors.accent,
        outline: `2px solid ${theme.colors.primary}`,
      },
    }),
  })
);
