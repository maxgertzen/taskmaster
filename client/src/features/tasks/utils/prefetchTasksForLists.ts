import { QueryClient } from '@tanstack/react-query';

import { QUERY_KEYS } from '@/shared/api/query-keys';
import { STALE_TIME } from '@/shared/constants/staleTime';

import { fetchTasks } from '../api';

export const prefetchTasksForLists = (
  listIds: string[],
  queryClient: QueryClient,
  token: string | null
) => {
  if (!token || !listIds.length) return;

  listIds.forEach((listId) => {
    queryClient.prefetchQuery({
      queryKey: QUERY_KEYS.tasks({ listId }),
      queryFn: fetchTasks(token),
      staleTime: STALE_TIME,
    });
  });
};
