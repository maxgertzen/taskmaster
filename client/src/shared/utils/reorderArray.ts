export const reorderArray = <T>(
  array: T[],
  fromIndex: number,
  toIndex: number
): T[] => {
  const reordered = [...array];
  const [movedItem] = reordered.splice(fromIndex, 1);
  reordered.splice(toIndex, 0, movedItem);
  return reordered;
};
