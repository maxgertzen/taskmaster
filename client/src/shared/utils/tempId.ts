export const generateTempId = (prefix: string = 'temp') =>
  `${prefix}-${Date.now()}`;
