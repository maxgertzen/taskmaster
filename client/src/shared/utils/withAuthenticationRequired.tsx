import { withAuthenticationRequired as realWithAuthenticationRequired } from '@auth0/auth0-react';

import { IS_AUTH0_DISABLED } from '@/constants';
import { mockWithAuthenticationRequired } from '@/mocks';

const isMock = IS_AUTH0_DISABLED === 'true';

export const withAuthenticationRequired = isMock
  ? mockWithAuthenticationRequired
  : realWithAuthenticationRequired;
