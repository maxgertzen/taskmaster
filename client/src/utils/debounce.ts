export function debounce<T extends (...args: Parameters<T>) => ReturnType<T>>(
  func: T,
  delay: number
): ((...args: Parameters<T>) => void) & { cancel: () => void } {
  let timeoutId: ReturnType<typeof setTimeout> | null = null;

  const debouncedFunction = (...args: Parameters<T>) => {
    if (timeoutId) clearTimeout(timeoutId);
    timeoutId = setTimeout(() => func(...args), delay);
  };

  debouncedFunction.cancel = () => {
    if (timeoutId) clearTimeout(timeoutId);
  };

  return debouncedFunction;
}
