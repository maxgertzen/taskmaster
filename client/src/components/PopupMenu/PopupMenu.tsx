import { FC, ReactNode, useEffect, useRef, useState } from 'react';

import { MenuContainer, MenuItem } from './PopupMenu.styled';

interface PopupMenuProps {
  options: Array<{
    label: ReactNode;
    onClick?: () => void;
    disabled?: boolean;
  }>;
  onClose: () => void;
  isOpen: boolean;
  triggerRef: React.RefObject<HTMLElement>;
}

export const PopupMenu: FC<PopupMenuProps> = ({
  options,
  onClose,
  isOpen,
  triggerRef,
}) => {
  const menuRef = useRef<HTMLDivElement>(null);
  const [menuPosition, setMenuPosition] = useState<{
    top: number;
    left: number;
  }>({ top: 0, left: 0 });

  useEffect(() => {
    if (isOpen && triggerRef.current) {
      const triggerRect = triggerRef.current.getBoundingClientRect();
      setMenuPosition({
        top: triggerRect.bottom + window.scrollY,
        left: triggerRect.left + window.scrollX,
      });
    }
  }, [isOpen, triggerRef]);

  useEffect(() => {
    const handleClickOutside = (event: MouseEvent) => {
      if (
        menuRef.current &&
        !menuRef.current.contains(event.target as Node) &&
        triggerRef.current &&
        !triggerRef.current.contains(event.target as Node)
      ) {
        onClose();
      }
    };

    if (isOpen) {
      document.addEventListener('mouseup', handleClickOutside);
    }
    return () => {
      document.removeEventListener('mouseup', handleClickOutside);
    };
  }, [isOpen, onClose, triggerRef]);

  if (!isOpen) return null;

  return (
    <MenuContainer
      ref={menuRef}
      role='menu'
      aria-labelledby='popup-menu-trigger'
      top={menuPosition.top}
      onClick={(event) => event.stopPropagation()}
    >
      {options.map((option, index) => (
        <MenuItem
          key={index}
          role='menuitem'
          onClick={() => {
            if (!option.disabled) {
              option?.onClick?.();
              onClose();
            }
          }}
          disabled={option.disabled}
        >
          {option.label}
        </MenuItem>
      ))}
    </MenuContainer>
  );
};
