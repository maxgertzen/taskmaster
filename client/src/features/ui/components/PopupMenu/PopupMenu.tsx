import { FC, ReactNode, useEffect, useRef } from 'react';

import { MenuContainer, MenuItem } from './PopupMenu.styled';

interface PopupMenuProps {
  options: Array<{
    label: ReactNode;
    onClick?: () => void;
    disabled?: boolean;
  }>;
  onClose: () => void;
  isOpen: boolean;
  orientation?: 'left' | 'right' | 'full';
}

export const PopupMenu: FC<PopupMenuProps> = ({
  options,
  onClose,
  isOpen,
  orientation = 'left',
}) => {
  const menuRef = useRef<HTMLDivElement>(null);

  useEffect(() => {
    const handleClickOutside = (event: MouseEvent) => {
      if (menuRef.current && !menuRef.current.contains(event.target as Node)) {
        onClose();
      }
    };

    if (isOpen) {
      document.addEventListener('mouseup', handleClickOutside);
    }
    return () => {
      document.removeEventListener('mouseup', handleClickOutside);
    };
  }, [isOpen, onClose]);

  if (!isOpen) return null;

  return (
    <MenuContainer
      ref={menuRef}
      data-testid='popup-menu'
      role='menu'
      aria-labelledby='popup-menu-trigger'
      orientation={orientation}
      onClick={(event) => event.stopPropagation()}
    >
      {options.map((option, index) => (
        <MenuItem
          key={index}
          data-testid='popup-menu-item'
          role='menuitem'
          onClick={() => {
            if (!option.disabled) {
              option?.onClick?.();
              onClose();
            }
          }}
          disabled={option.disabled}
        >
          {option.label}
        </MenuItem>
      ))}
    </MenuContainer>
  );
};
