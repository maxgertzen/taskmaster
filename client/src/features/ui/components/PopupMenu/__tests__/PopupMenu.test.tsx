import { fireEvent } from '@testing-library/dom';
import { describe, it, expect, vi, beforeEach } from 'vitest';

import { render, screen } from '@/tests/utils';

import { PopupMenu } from '../PopupMenu';

describe('PopupMenu', () => {
  const mockOnClose = vi.fn();
  const mockOptionClick = vi.fn();

  const defaultOptions = [
    { label: 'Option 1', onClick: mockOptionClick },
    { label: 'Option 2', onClick: mockOptionClick },
  ];

  beforeEach(() => {
    vi.clearAllMocks();
  });

  describe.each([['left'], ['right'], ['full']])(
    'with %s orientation',
    (orientation) => {
      it.each([
        ['light', undefined],
        ['dark', 'dark'],
      ])('renders correctly in %s theme when open', (_, theme) => {
        const { container } = render(
          <PopupMenu
            options={defaultOptions}
            onClose={mockOnClose}
            isOpen={true}
            orientation={orientation as 'left' | 'right' | 'full'}
          />,
          { theme: theme as 'light' | 'dark' }
        );
        expect(container).toMatchSnapshot();
      });
    }
  );

  describe('Visibility', () => {
    it('renders nothing when closed', () => {
      render(
        <PopupMenu
          options={defaultOptions}
          onClose={mockOnClose}
          isOpen={false}
        />
      );
      expect(screen.queryByTestId('popup-menu')).not.toBeInTheDocument();
    });

    it('renders menu when open', () => {
      render(
        <PopupMenu
          options={defaultOptions}
          onClose={mockOnClose}
          isOpen={true}
        />
      );
      expect(screen.queryByTestId('popup-menu')).toBeInTheDocument();
    });
  });

  describe('User Interactions', () => {
    it('calls option onClick and closes when option is clicked', async () => {
      const { user } = render(
        <PopupMenu
          options={defaultOptions}
          onClose={mockOnClose}
          isOpen={true}
        />
      );

      await user.click(screen.getByText('Option 1'));
      expect(mockOptionClick).toHaveBeenCalledTimes(1);
      expect(mockOnClose).toHaveBeenCalledTimes(1);
    });

    it('handles disabled options', async () => {
      const options = [
        { label: 'Disabled Option', onClick: mockOptionClick, disabled: true },
      ];

      const { user } = render(
        <PopupMenu options={options} onClose={mockOnClose} isOpen={true} />
      );

      await user.click(screen.getByText('Disabled Option'));
      expect(mockOptionClick).not.toHaveBeenCalled();
      expect(mockOnClose).not.toHaveBeenCalled();
    });

    it('closes when clicking outside', () => {
      render(
        <PopupMenu
          options={defaultOptions}
          onClose={mockOnClose}
          isOpen={true}
        />
      );

      fireEvent.mouseUp(document);
      expect(mockOnClose).toHaveBeenCalled();
    });

    it('does not close when clicking inside menu', () => {
      render(
        <PopupMenu
          options={defaultOptions}
          onClose={mockOnClose}
          isOpen={true}
        />
      );

      const menu = screen.queryByTestId('popup-menu');
      if (!menu) return;

      fireEvent.click(menu);
      expect(mockOnClose).not.toHaveBeenCalled();
    });
  });
});
