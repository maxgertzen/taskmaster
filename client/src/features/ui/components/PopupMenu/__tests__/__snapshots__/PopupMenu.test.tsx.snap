// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PopupMenu > with full orientation > renders correctly in dark theme when open 1`] = `
<div>
  <div
    aria-labelledby="popup-menu-trigger"
    class="css-1g1q5t1"
    data-testid="popup-menu"
    orientation="full"
    role="menu"
  >
    <span
      class="css-vjep5o"
      data-testid="popup-menu-item"
      role="menuitem"
    >
      Option 1
    </span>
    <span
      class="css-vjep5o"
      data-testid="popup-menu-item"
      role="menuitem"
    >
      Option 2
    </span>
  </div>
</div>
`;

exports[`PopupMenu > with full orientation > renders correctly in light theme when open 1`] = `
<div>
  <div
    aria-labelledby="popup-menu-trigger"
    class="css-ufn8v6"
    data-testid="popup-menu"
    orientation="full"
    role="menu"
  >
    <span
      class="css-t8ojuo"
      data-testid="popup-menu-item"
      role="menuitem"
    >
      Option 1
    </span>
    <span
      class="css-t8ojuo"
      data-testid="popup-menu-item"
      role="menuitem"
    >
      Option 2
    </span>
  </div>
</div>
`;

exports[`PopupMenu > with left orientation > renders correctly in dark theme when open 1`] = `
<div>
  <div
    aria-labelledby="popup-menu-trigger"
    class="css-19la4k7"
    data-testid="popup-menu"
    orientation="left"
    role="menu"
  >
    <span
      class="css-vjep5o"
      data-testid="popup-menu-item"
      role="menuitem"
    >
      Option 1
    </span>
    <span
      class="css-vjep5o"
      data-testid="popup-menu-item"
      role="menuitem"
    >
      Option 2
    </span>
  </div>
</div>
`;

exports[`PopupMenu > with left orientation > renders correctly in light theme when open 1`] = `
<div>
  <div
    aria-labelledby="popup-menu-trigger"
    class="css-mr4zrh"
    data-testid="popup-menu"
    orientation="left"
    role="menu"
  >
    <span
      class="css-t8ojuo"
      data-testid="popup-menu-item"
      role="menuitem"
    >
      Option 1
    </span>
    <span
      class="css-t8ojuo"
      data-testid="popup-menu-item"
      role="menuitem"
    >
      Option 2
    </span>
  </div>
</div>
`;

exports[`PopupMenu > with right orientation > renders correctly in dark theme when open 1`] = `
<div>
  <div
    aria-labelledby="popup-menu-trigger"
    class="css-qw809l"
    data-testid="popup-menu"
    orientation="right"
    role="menu"
  >
    <span
      class="css-vjep5o"
      data-testid="popup-menu-item"
      role="menuitem"
    >
      Option 1
    </span>
    <span
      class="css-vjep5o"
      data-testid="popup-menu-item"
      role="menuitem"
    >
      Option 2
    </span>
  </div>
</div>
`;

exports[`PopupMenu > with right orientation > renders correctly in light theme when open 1`] = `
<div>
  <div
    aria-labelledby="popup-menu-trigger"
    class="css-f0tv5q"
    data-testid="popup-menu"
    orientation="right"
    role="menu"
  >
    <span
      class="css-t8ojuo"
      data-testid="popup-menu-item"
      role="menuitem"
    >
      Option 1
    </span>
    <span
      class="css-t8ojuo"
      data-testid="popup-menu-item"
      role="menuitem"
    >
      Option 2
    </span>
  </div>
</div>
`;
